Added API for getAllPlugins

alex-parisi opened the pull request:
Added functionality to the Scripting API to allow Contexts to return a list of all registered plugins. Will be useful for designing a plugin-manager plugin.

To use, call `context.allPlugins()` in a plugin. This returns an object that contains useful metadata from each registered plugin. Additional metadata parameters can be added later if needed.

CookiePLMonster commented on src/openrct2/scripting/bindings/game/ScContext.hpp:
```diff
@@ -45,6 +45,36 @@ namespace OpenRCT2::Scripting
+            auto ctx = scriptEngine.GetContext();
+            std::vector<DukValue> result;
+            // Iterate through all plugins and and cast their data to Duk objects
+            for (std::shared_ptr<OpenRCT2::Scripting::Plugin> plugin_ptr : allPlugins)
```
```suggestion
            for (const std::shared_ptr<OpenRCT2::Scripting::Plugin>& plugin_ptr : allPlugins)
```

It'll probably make the line below fail to compile, but adding constness to it too might help.

ZehMatt commented on src/openrct2/scripting/bindings/game/ScContext.hpp:
```diff
@@ -45,6 +45,36 @@ namespace OpenRCT2::Scripting
+        {
+            // Get all of the plugins from the script engine
+            OpenRCT2::Scripting::ScriptEngine& scriptEngine = GetContext()->GetScriptEngine();
+            std::vector<std::shared_ptr<OpenRCT2::Scripting::Plugin>> allPlugins = scriptEngine.GetPlugins();
```
```suggestion
            const auto& allPlugins = scriptEngine.GetPlugins();
```
Making a copy of this is quite expensive especially with the shared_ptrs being the elements.

alex-parisi replied:
Thanks for this little explanation, helped a lot with my understanding of references

ZehMatt commented on src/openrct2/scripting/bindings/game/ScContext.hpp:
```diff
@@ -45,6 +45,36 @@ namespace OpenRCT2::Scripting
+                dObj.Set("name", metadata.Name);
+                dObj.Set("version", metadata.Version);
+                // Concatenate all authors into one string
+                std::string authors = "";
+                for (std::string s : metadata.Authors)
+                {
+                    authors += s + " ";
+                }
+
+                dObj.Set("authors", authors);
```
Would be better to return an array of authors, plugins can decide how to deal with displaying this data. In javascript this is quite easy with a join call

alex-parisi replied:
Definitely is better to return the array of authors instead, just took some playing around to get it to work. Before, I was using the DukObj class to encapsulate the PluginMetadata, but that didn't have a Set for string arrays. I'm now just interacting directly with the Duk stack to cast the metadata now.

ZehMatt commented on distribution/openrct2.d.ts:
```diff
@@ -172,6 +172,12 @@ declare global {
+        /**
+         * Gets all of the currently registered plugins
+         */
+        allPlugins(): PluginMetadata[];
```
```suggestion
        getPlugins(): PluginMetadata[];
```
Most of the scripting API has get/set

ZehMatt commented on src/openrct2/scripting/bindings/game/ScContext.hpp:
```diff
@@ -45,6 +45,36 @@ namespace OpenRCT2::Scripting
         }
 
     private:
+        std::vector<DukValue> allPlugins_get()
```
```suggestion
        std::vector<DukValue> getPlugins()
```
The get/set postfix should be used for properties and not methods.

alex-parisi replied:
I understand, will use this convention from now on. Fixed in the latest commit.

alex-parisi commented:
The latest commit fixes these issues:

1. Adopted correct get/set naming conventions
2. Removed unnecessary copies, using `auto` more often
3. Changed the way authors are handled: instead of concatenating all authors into a string (lame!) we can now return an array of strings for the authors instead. This required understanding how a `DukObj` interacted with the Duk stack, since there is no `Set` method for a string array. 

Note that instead of manually interacting with the Duk stack to create the metadata object, we could implement a `Set` method for the `DukObj` class to handle string arrays easily.

Here is an example of a plugin using the new API:
![example_plugin](https://github.com/OpenRCT2/OpenRCT2/assets/6962601/2553dda9-7596-4fb8-8bbd-c1fc21c89fa9)

ZehMatt commented:
The changes seem fine, this requires the plugin API to be raised

alex-parisi commented:
Increased the Plugin API version from `78` to `79` in commit [9eede8b](https://github.com/OpenRCT2/OpenRCT2/pull/20709/commits/9eede8b75429f91343db17944b5e7975e24399e6)

ZehMatt commented:
Sorry I forgot this needs also a changelog entry

alex-parisi commented:
No worries, I added an entry in the changelog in the latest commit

ZehMatt commented on distribution/changelog.txt:
```diff
@@ -1,5 +1,6 @@
 0.4.6 (in development)
 ------------------------------------------------------------------------
+- Feature: [Plugin] Added API for returning metadata from all registered plugins.
```
We typically have the entry for plugins like this.
`- Feature: [#18675] [Plugin] Plugins can refer to g2 image icons by name.`
You can use your PR number

alex-parisi replied:
Got it, I added the PR number and re-worded it a bit

ZehMatt commented on distribution/changelog.txt:
```diff
@@ -1,6 +1,6 @@
 0.4.6 (in development)
 ------------------------------------------------------------------------
-- Feature: [Plugin] Added API for returning metadata from all registered plugins.
+- Feature: [#20709] [Plugin] Plugins can now check metadata from all registered plugins.
```
We sort the changelog entries by the issue/PR number, please move it to the correct line.

alex-parisi replied:
Understood, my bad. Moved it to the correct location in the latest commit.

IntelOrca commented:
I think it makes sense to avoid poluting `Context` more than it already is and instead create a new interface: `PluginManager`:

```ts
global {
    var pluginManager: PluginManager;
}

interface PluginManager {
    plugins: PluginMetadata[];
}
```

ZehMatt commented:
> I think it makes sense to avoid poluting `Context` more than it already is and instead create a new interface: `PluginManager`:
> 
> ```ts
> global {
>     var pluginManager: PluginManager;
> }
> 
> interface PluginManager {
>     plugins: PluginMetadata[];
> }
> ```
> 
> Because you will need more functionality at some point and this keeps plugin related APIs together.

What functionality would be added later on?

IntelOrca commented:
> What functionality would be added later on?

Currently all I see in this is a way to show the details of each one. There are not yet APIs to enable or disable them, install new ones, or debug them, restart them etc.

ZehMatt commented:
> > What functionality would be added later on?
> 
> Currently all I see in this is a way to show the details of each one. There are not yet APIs to enable or disable them, install new ones, or debug them, restart them etc.

I don't think plugins should get access to things like installing or removing other plugins, I don't mind having the manager written in JS but I think that should be in an isolated context and we could have that included by default, imagine joining a server that queries all plugins and just deletes them, don't think that's a good idea to have. Debugging API is a good idea but I'm not sure how that fits into the PluginManager interface.

IntelOrca commented:
I am trying to be consistent with other things. ObjectManager, TitleSequenceManager, Profiler, have their APIs grouped behind a manager interface.

ZehMatt commented:
> I am trying to be consistent with other things. ObjectManager, TitleSequenceManager, Profiler, ImageManager have their APIs grouped behind a manager interface.

I think for the Debugging part it best to have it behind a Debugging/DebuggingManager interface. Anyway getting a bit off-topic here.

alex-parisi commented:
I agree with @IntelOrca, makes sense to put it all into its own interface if other features are added later. The idea was to implement disabling/enabling next. Plus, wrapping it up in its own class is easy enough to implement. 

@ZehMatt, you make a good point about servers uninstalling all plugins (or even installing malicious ones) but I do think there could be a use for a plugin manager as long as its scope is limited.

ZehMatt commented:
I'm fine with the change of moving it into a new interface, @IntelOrca should this be a read-only property to get all plugins or is the method acceptable?

IntelOrca commented:
Other similar interfaces have used a property. Methods tend to be only used if parameters are required, such as `getRide`.

I was thinking for plugin managers, they could use a new plugin scope which gives them access to the mutable APIs. And uninstalling / installing plugins could certainly be limited to the title sequence, or client only.

alex-parisi commented:
The only issue with changing it to be a read-only property and not a method is that if the list of registered plugins changes (by uninstalling or installing new ones if hot-reloading is enabled), then the property won't be up-to-date. Right now it behaves similarly to `context.getAllTrackSegments()`, which is a method that takes no inputs. Although I don't have much experience with TypeScript interfaces so I could be wrong.

IntelOrca commented:
> The only issue with changing it to be a read-only property and not a method is that if the list of registered plugins changes (by uninstalling or installing new ones if hot-reloading is enabled), then the property won't be up-to-date. Right now it behaves similarly to `context.getAllTrackSegments()`, which is a method that takes no inputs. Although I don't have much experience with TypeScript interfaces so I could be wrong.

The property should generate the value when it is called. This is how the rides property works, it is dynamic.

alex-parisi commented:
Makes sense, thanks for explaining it! I changed it from a method to a property in the latest commit

duncanspumpkin commented:
@alex-parisi could you rebase this?

tupaschoal commented:
I've rebased this and fixed the merge issues, also bumped the plugin version.
Seems like it was mostly ready, but fell through the cracks. Can you review @IntelOrca and @ZehMatt ?

ZehMatt commented on distribution/openrct2.d.ts:
```diff
@@ -4948,4 +4951,11 @@ declare global {
+     * Interface to handle the plugin manager
+     */
+    interface PluginManager {
+        readonly plugins: PluginMetadata[];
```
Where is the PluginMetadata type defined? I think if we merge this PR we should not have this missing.

tupaschoal replied:
It existed already, isn't it this:

https://github.com/OpenRCT2/OpenRCT2/blob/f10911194507a88ca3e05d8f1f35b738d27ae353/distribution/openrct2.d.ts#L140-L153

ZehMatt replied:
Didn't realize its already there, nevermind then.
